### PR TITLE
Add test fqdn_valid

### PR DESCRIPTION
**Pull request by vbotka**
##### SUMMARY

Test validity of a hostname. See https://github.com/ansible/ansible/issues/81831. This test provides systemic testing of hostnames instead of the required update of the module *hostname*.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
fqdn_valid

##### ADDITIONAL INFORMATION
The test requires PyPI package *fqdn*. A playbook can easily test the validity of the hostnames

```yaml
    - assert:
        that: hosts_invalid|length == 0
        fail_msg: "[ERR] Invalid FQDN: {{ hosts_invalid }}"
      vars:
        hosts_invalid: "{{ ansible_play_hosts_all|reject('fqdn_valid') }}"
      run_once: true
```

```
fatal: [node1]: FAILED! => changed=false 
  assertion: hosts_invalid|length == 0
  evaluated_to: false
  msg: '[ERR] Invalid FQDN: [''-node3'']'
```

### CODE REVIEW

File: changelogs/fragments/7533-add-test-fqdn_valid.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - fqdn_valid test - adds test to validates FQDN names against RFC 112 (https://github.com/ansible-collections/community.general/pull/7533).
```
Comment (felixfontein) [inline changelogs/fragments/7533-add-test-fqdn_valid.yml:2]: Please remove this, new tests are documented by their `version_added` + `short_description` in documentation.

File: plugins/test/fqdn_valid.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+  name: fqdn_valid
+  short_description: Validates fully-qualified domain names against RFC 1123.
+  version_added: 8.0.3
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:11]:
```suggestion
  version_added: 8.1.0
```

File: plugins/test/fqdn_valid.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    that: hostname is community.general.fqdn_valid(min_labels=2, allow_underscores=True)
+'''
+
+RETURN = '''
+  _value:
+    description: Whether the name is valid.
+    type: bool
+'''
+
+from fqdn import FQDN
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:66]: This import needs to be guarded (see https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/import.html).

File: plugins/test/fqdn_valid.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+try:
+    from ansible.errors import AnsiblePluginRemovedError
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:69]: Not sure, but what is this for?

File: plugins/test/fqdn_valid.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      RFC-conforming software, which is a strict subset of the logic
+      in modern web browsers like Mozilla Firefox and Chromium that
+      determines whether make a DNS lookup.
+    - Certificate Authorities like Let's Encrypt run a narrower set of
+      string validation logic to determine validity for issuance. This
+      test is not intended to achieve functional parity with CA
+      issuance
+    - Single label names are allowed by default (min_labels=1).
+  options:
+    name:
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:30]: The test's input should be named `_input`:
```suggestion
    _input:
```

File: plugins/test/fqdn_valid.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    - The design intent is to validate that a string would be
+      traditionally acceptable as a public Internet hostname to
+      RFC-conforming software, which is a strict subset of the logic
+      in modern web browsers like Mozilla Firefox and Chromium that
+      determines whether make a DNS lookup.
+    - Certificate Authorities like Let's Encrypt run a narrower set of
+      string validation logic to determine validity for issuance. This
+      test is not intended to achieve functional parity with CA
+      issuance
+    - Single label names are allowed by default (min_labels=1).
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:28]:
```suggestion
    - Single label names are allowed by default (O(min_labels=1)).
```

File: plugins/test/fqdn_valid.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      type: str
+      required: true
+    min_labels:
+      description: Required minimum of labels, separated by period
+      default: 1
+      type: int
+      required: false
+    allow_underscores:
+      description: Allow underscore characters.
+      default: False
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:41]:
```suggestion
      default: false
```

File: tests/integration/targets/test_fqdn_valid/aliases [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) Ansible Project
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+azp/posix/2
+skip/python2.6  # filters are controller only, and we no longer support Python 2.6 on the controller
```
Comment (felixfontein) [inline tests/integration/targets/test_fqdn_valid/aliases:6]: We just got rid of Python 2.6 support ;) (Yes, I know, this exists in more places, we'll clean them up eventually.)
```suggestion
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

File: tests/integration/targets/test_fqdn_valid/aliases (hunk 8 above)
Comment (felixfontein) [reply tests/integration/targets/test_fqdn_valid/aliases:6]: #7534

File: plugins/test/fqdn_valid.py (hunk 4 above)
Comment (vbotka) [reply plugins/test/fqdn_valid.py:77]: Neither am I. There is only one test in *community.general* 
test/a_module.py

I took it from there.

Comment (felixfontein) [reply plugins/test/fqdn_valid.py:77]: That other test explicitly needs this exception. This test does not, so this import isn't needed.

**Comment by vbotka** [comment]
I don't know how to resolve the below error reported by Azure Pipelines CI Sanity

```
ERROR: plugins/test/fqdn_valid.py:92:11: E275: missing whitespace after keyword (100%)
```
```python
   return(fobj.is_valid)
          ^
          92:11
```
I'm not able to reproduce the problem with *pycodestyle* 2.5.0. (Ignoring E501 line too long).

```yaml
shell> python3.8 -m pycodestyle --verbose --ignore=E501 fqdn_valid.py 
checking fqdn_valid.py
```

File: plugins/test/fqdn_valid.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    from fqdn import FQDN
+except ImportError as imp_exc:
+    ANOTHER_LIBRARY_IMPORT_ERROR = imp_exc
+else:
+    ANOTHER_LIBRARY_IMPORT_ERROR = None
+
+
+DOCUMENTATION = '''
+  name: fqdn_valid
+  short_description: Validates fully-qualified domain names against RFC 1123.
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:21]:
```suggestion
  short_description: Validates fully-qualified domain names against RFC 1123
```

File: plugins/test/fqdn_valid.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  version_added: 8.1.0
+  author: Vladimir Botka (@vbotka)
+  requirements:
+  - fqdn>=1.5.1 (PyPI)
+  description:
+    - This test validates Fully Qualified Domain Names (FQDNs)
+      conforming to the Internet Engineering Task Force specification
+      RFC 1123 and RFC 952
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:29]:
```suggestion
      RFC 1123 and RFC 952.
```
Or let's better link the RFCs:
```suggestion
      L(RFC 1123, https://www.rfc-editor.org/rfc/rfc1123) and L(RFC 952, https://www.rfc-editor.org/rfc/rfc952).
```

File: plugins/test/fqdn_valid.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    - The design intent is to validate that a string would be
+      traditionally acceptable as a public Internet hostname to
+      RFC-conforming software, which is a strict subset of the logic
+      in modern web browsers like Mozilla Firefox and Chromium that
+      determines whether make a DNS lookup.
+    - Certificate Authorities like Let's Encrypt run a narrower set of
+      string validation logic to determine validity for issuance. This
+      test is not intended to achieve functional parity with CA
+      issuance
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:38]:
```suggestion
      issuance.
```

File: plugins/test/fqdn_valid.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    - Single label names are allowed by default (O(min_labels=1)).
+  options:
+    _input:
+      description: Name of the host.
+      type: str
+      required: true
+    min_labels:
+      description: Required minimum of labels, separated by period
```
Comment (felixfontein) [inline plugins/test/fqdn_valid.py:46]:
```suggestion
      description: Required minimum of labels, separated by period.
```

**Review by felixfontein — COMMENTED** [review]
Just some small docs improvements, besides that LGTM!

**Review by felixfontein — APPROVED** [review]
LGTM! If nobody objects, I'll merge this in a few days.

**Comment by felixfontein** [comment]
@vbotka thanks for your contribution!

**Pull request merged**